apps/create-fuels-counter-guide/src/pages/predicate.tsx

```diff
@@ -0,0 +1,180 @@
+import { BN, InputValue, Predicate } from "fuels";
+import { bn } from "fuels";
+import { useState } from "react";
+import toast from "react-hot-toast";
+import useAsync from "react-use/lib/useAsync";
+
+export default function PredicateExample() {
+  let baseAssetId: string;
+
+  const { wallet, walletBalance, refreshWalletBalance } = useActiveWallet();
+
+  const [predicate, setPredicate] = useState<Predicate<InputValue[]>>();
+
+  const [predicateBalance, setPredicateBalance] = useState<BN>();
+
+  const [pin, setPin] = useState<string>();
+
+  useAsync(async () => {
+    if (wallet) {
+      baseAssetId = wallet.provider.getBaseAssetId();
+      const predicate = TestPredicateAbi__factory.createInstance(wallet.provider);
+      setPredicate(predicate);
+      setPredicateBalance(await predicate.getBalance());
+    }
+  }, [wallet]);
+
+  const refreshBalances = async () => {
+    await refreshWalletBalance?.();
+    setPredicateBalance(await predicate?.getBalance());
+  };
+
+  const transferFundsToPredicate = async (amount: BN) => {
+    if (!predicate) {
+      return toast.error("Predicate not loaded");
+    }
+
+    if (!wallet) {
+      return toast.error("Wallet not loaded");
+    }
+
+    await wallet.transfer(predicate.address, amount, baseAssetId, {
+      gasLimit: 10_000,
+    });
+
+    await refreshBalances();
+
+    return toast.success("Funds transferred to predicate.");
+  };
+
+  const unlockPredicateAndTransferFundsBack = async (amount: BN) => {
+    try {
+      if (!wallet) {
+        return toast.error("Wallet not loaded");
+      }
+
+      const reInitializePredicate = TestPredicateAbi__factory.createInstance(wallet.provider, [bn(pin)]);
+
+      if (!reInitializePredicate) {
+        return toast.error("Failed to initialize predicate");
+      }
+
+      const tx = await reInitializePredicate.transfer(wallet.address, amount, baseAssetId);
+      const { isStatusSuccess } = await tx.wait();
+
+      if (!isStatusSuccess) {
+        toast.error("Failed to unlock predicate");
+      }
+
+      if (isStatusSuccess) {
+        toast.success("Predicate unlocked");
+      }
+
+      await refreshBalances();
+    } catch (e) {
+      console.error(e);
+      toast.error(
+        "Failed to unlock predicate. You probably entered the wrong pin, or the predicate does not have enough balance. Try again."
+      );
+    }
+  };
+
+  // #region change-pin-react-function
+  const changePin = async () => {
+    if (!wallet) {
+      return toast.error("Wallet not loaded");
+    }
+    if (!predicate) {
+      return toast.error("Predicate not loaded");
+    }
+
+    if (walletBalance?.eq(0)) {
+      return toast.error(
+        "Your wallet does not have enough funds. Please click the 'Top-up Wallet' button in the top right corner, or use the local faucet."
+      );
+    }
+
+    if (!pin) {
+      return toast.error("Please enter a pin");
+    }
+
+    const configurable = { PIN: bn(pin) };
```

Comment:
@Dhaiwat10 I'm uncertain as to whether this is idiomatic React, I primarily did things this way as it seemed to be more favourable to the nuances of how React manages state. I'd be keen to hear your insights.